File: src/main/scala/com/yogpc/qp/packet/enchantment/MoverMessage.java
package com.yogpc.qp.packet.enchantment;

import java.io.IOException;

import com.yogpc.qp.container.ContainerMover;
import com.yogpc.qp.container.ContainerMover.D;
import com.yogpc.qp.packet.IMessage;
import net.minecraft.inventory.Container;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public class MoverMessage {
    /**
     * To server only.
     * For container player opening.
     */
    public static class Move implements IMessage {

        BlockPos pos;
        int id;

        public static Move create(BlockPos pos, int id) {
            Move move = new Move();
            move.pos = pos;
            move.id = id;
            return move;
        }

        @Override
        public void fromBytes(PacketBuffer buffer) throws IOException {
            pos = buffer.readBlockPos();
            id = buffer.readInt();
        }

        @Override
        public void toBytes(PacketBuffer buffer) {
            buffer.writeBlockPos(pos).writeInt(id);
        }

        @Override
        public IMessage onRecieve(IMessage message, MessageContext ctx) {
            Container container = ctx.getServerHandler().playerEntity.openContainer;
            if (container.windowId == id) {
                ((ContainerMover) container).moveEnchant();
            }
            return null;
        }
    }

    /**
     * To server only.
     * For container player opening.
     */
    public static class Cursor implements IMessage {
        D d;
        int id;
        BlockPos pos;

        public static Cursor create(BlockPos pos, int id, D d) {
            Cursor cursor = new Cursor();
            cursor.d = d;
            cursor.id = id;
            cursor.pos = pos;
            return cursor;
        }

        @Override
        public void fromBytes(PacketBuffer buffer) throws IOException {
            d = buffer.readEnumValue(D.class);
            pos = buffer.readBlockPos();
            id = buffer.readInt();
        }

        @Override
        public void toBytes(PacketBuffer buffer) {
            buffer.writeEnumValue(d).writeBlockPos(pos).writeInt(id);
        }

        @Override
        public IMessage onRecieve(IMessage message, MessageContext ctx) {
            Container container = ctx.getServerHandler().playerEntity.openContainer;
            if (container.windowId == id) {
                ((ContainerMover) container).setAvail(d);
            }
            return null;
        }
    }

}

File: src/main/scala/com/yogpc/qp/packet/enchantment/MoverMessage.scala
package com.yogpc.qp.packet.enchantment

import java.io.IOException

import com.yogpc.qp.QuarryPlus
import com.yogpc.qp.container.ContainerMover
import com.yogpc.qp.packet.IMessage
import net.minecraft.network.PacketBuffer
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext

object MoverMessage {
    def create(pos: BlockPos, id: Int): Move = {
        val move = new Move(pos, id)
        move
    }

    def create(pos: BlockPos, id: Int, d: ContainerMover.D): Cursor = {
        val cursor = new Cursor(pos, id, d)
        cursor
    }

    /**
      * To server only.
      * For container player opening.
      */
    class Move(var pos: BlockPos, var id: Int) extends IMessage {

        def this() {
            this(null, 0)
        }

        @throws[IOException]
        override def fromBytes(buffer: PacketBuffer): Unit = {
            pos = buffer.readBlockPos
            id = buffer.readInt
        }

        override def toBytes(buffer: PacketBuffer): Unit = buffer.writeBlockPos(pos).writeInt(id)

        override def onRecieve(message: IMessage, ctx: MessageContext): IMessage = {
            val server = QuarryPlus.proxy.getPacketWorld(ctx.netHandler).getMinecraftServer
            Option(server).foreach(_.addScheduledTask(IMessage.toRunnable(() => {
                val container = QuarryPlus.proxy.getPacketPlayer(ctx.netHandler).openContainer
                if (container.windowId == id) container.asInstanceOf[ContainerMover].moveEnchant()
            })))
            null
        }
    }

    /**
      * To server only.
      * For container player opening.
      */
    class Cursor(var pos: BlockPos, var id: Int, var d: ContainerMover.D) extends IMessage {

        def this() {
            this(null, 0, null)
        }

        @throws[IOException]
        override def fromBytes(buffer: PacketBuffer): Unit = {
            d = buffer.readEnumValue(classOf[ContainerMover.D])
            pos = buffer.readBlockPos
            id = buffer.readInt
        }

        override def toBytes(buffer: PacketBuffer): Unit = buffer.writeEnumValue(d).writeBlockPos(pos).writeInt(id)

        override def onRecieve(message: IMessage, ctx: MessageContext): IMessage = {
            val server = QuarryPlus.proxy.getPacketWorld(ctx.netHandler).getMinecraftServer
            Option(server).foreach(_.addScheduledTask(IMessage.toRunnable(() => {
                val container = QuarryPlus.proxy.getPacketPlayer(ctx.netHandler).openContainer
                if (container.windowId == id) container.asInstanceOf[ContainerMover].setAvail(d)
            })))
            null
        }
    }

}
